proxyclient/client.go

```diff
@@ -0,0 +1,222 @@
+package proxyclient
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/rancher/remotedialer"
+	v1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	defaultServerAddr        = "wss://127.0.0.1"
+	defaultServerPort        = 5555
+	defaultServerPath        = "/connect"
+	retryTimeout             = 1 * time.Second
+	certificateWatchInterval = 10 * time.Second
+)
+
+type PortForwarder interface {
+	Start() error
+	Stop()
+}
+
+type ProxyClientOpt func(*ProxyClient)
+
+type ProxyClient struct {
+	forwarder           PortForwarder
+	serverUrl           string
+	serverConnectSecret string
+
+	dialer    *websocket.Dialer
+	dialerMtx sync.Mutex
+
+	secretController v1.SecretController
+	namespace        string
+	certSecretName   string
+	certServerName   string
+
+	onConnect func(ctx context.Context, session *remotedialer.Session) error
+}
+
+func New(ctx context.Context, serverSharedSecret, namespace, certSecretName, certServerName string, secretController v1.SecretController, forwarder PortForwarder, opts ...ProxyClientOpt) (*ProxyClient, error) {
+	if secretController == nil {
+		return nil, fmt.Errorf("SecretController required")
+	}
+
+	if forwarder == nil {
+		return nil, fmt.Errorf("a PortForwarder must be provided")
+	}
+
+	if namespace == "" {
+		return nil, fmt.Errorf("namespace required")
+	}
+
+	if certSecretName == "" {
+		return nil, fmt.Errorf("certSecretName required")
+	}
+
+	if serverSharedSecret == "" {
+		return nil, fmt.Errorf("server shared secret must be provided")
+	}
+
+	serverUrl := fmt.Sprintf("%s:%d%s", defaultServerAddr, defaultServerPort, defaultServerPath)
+
+	client := &ProxyClient{
+		serverUrl:           serverUrl,
+		forwarder:           forwarder,
+		serverConnectSecret: serverSharedSecret,
+		certSecretName:      certSecretName,
+		certServerName:      certServerName,
+		namespace:           namespace,
+	}
+
+	if err := client.buildDialer(ctx, secretController); err != nil {
+		return nil, fmt.Errorf("dialer build failed %w: ", err)
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	return client, nil
+}
+
+func (c *ProxyClient) buildDialer(ctx context.Context, secretController v1.SecretController) error {
+	secretController.OnChange(ctx, "remotedialer-proxy", func(_ string, newSecret *corev1.Secret) (*corev1.Secret, error) {
+		if newSecret.Name == c.certSecretName && newSecret.Namespace == c.namespace {
+			rootCAs, err := buildCertFromSecret(c.namespace, c.certSecretName, newSecret)
+			if err != nil {
+				logrus.Errorf("build certificate failed: %s", err.Error())
```

Comment:
nit: We're already returning the error so wrangler will also print it, meaning we'll print the error twice. 

I'd suggest just wrapping the error and returning it.